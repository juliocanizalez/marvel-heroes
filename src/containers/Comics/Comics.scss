@use '../../styles/global';

.comics{
  text-align: center;
  @extend %marvel-font;
  &-page__title{
    h1{
      font-size: 2.3rem;
      font-weight: global.$font-bold;
    }
  }
  &-search{
    form{
      display: flex;
      flex-direction: column;
      .comics-search__title{
        margin-top: 1.7rem;
        margin-bottom: 1.4rem;
      }
    }
  }
}

.search{
  display: flex;
  align-items: center;
  width: 100%;
  &-header{
    font-size: 1.3rem;
    margin-right: 1rem;
  }
  &-value{
    .search-input{
      outline: none;
      height: 1.9rem;
      padding: 0.2rem 1rem;
      @extend %marvel-font;
      font-size: 1.2rem;
      width: 250px;
    }
    &__select{
      margin-left: 1rem;
      height: 1.9rem;
      @extend %marvel-font;
      padding: 0.2rem 1rem;
      font-size: 1.1rem;
      width: 200px;
    }
  }
}

.cards{
  margin-top: 2rem;
  width: 100%;
  &-content{
    display: grid;
    gap: 1rem;
    grid-template-columns: repeat(5, 1fr);
  }
}