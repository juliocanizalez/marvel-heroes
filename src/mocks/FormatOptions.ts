interface IFormatOptions {
  name: string;
  value: string;
}
const FormatOptions = [
  {
    value: 'comic',
    name: 'Comic',
  },
  {
    value: 'magazine',
    name: 'Magazine',
  },
  {
    value: 'trade paperback',
    name: 'Trade Paperback',
  },
  {
    value: 'hardcover',
    name: 'Hardcover',
  },
  {
    value: 'digest',
    name: 'Digest',
  },
  {
    value: 'graphic novel',
    name: 'Graphic Novel',
  },
  {
    value: 'digital comic',
    name: 'Digital Comic',
  },
  {
    value: 'infinite comic',
    name: 'Infinite Comic',
  },
] as IFormatOptions[];

export default FormatOptions;
