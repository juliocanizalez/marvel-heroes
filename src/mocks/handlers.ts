/* eslint-disable import/extensions */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable arrow-body-style */
import { rest } from 'msw';
import {
  characterResponse,
  charactersResponse,
  comicResponse,
  comicsResponse,
  storiesResponse,
  storyResponse,
} from './testData';

const handlers = [
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(charactersResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters/:idCharacter`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(characterResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(comicsResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics/:idComic`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(comicResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(storiesResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories/:idStory`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(storyResponse()));
  }),
  rest.get('*', (req, res, ctx) => {
    // eslint-disable-next-line no-console
    console.error(`Please add request handler for ${req.url.toString()}`);
    return res(ctx.status(500), ctx.json({ error: 'Please add handler' }));
  }),
];

export default handlers;
