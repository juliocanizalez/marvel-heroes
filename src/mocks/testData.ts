/* eslint-disable */
import ICharacter from '../interfaces/ICharacter';
import IComic from '../interfaces/IComic';
import IStory from '../interfaces/IStory';

const createGenericApiResponse = <T>(results: T[]) => ({
  data: {
    offset: 0,
    limit: 12,
    total: 1,
    count: 1,
    results,
  },
});

const emptyResponse = () => ({
  data: {
    offset: 0,
    limit: 12,
    total: 0,
    count: 0,
    results: [],
  },
});

const charactersResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      comics: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
    },
  ]);

const characterResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
    },
  ]);

const comicsResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      characters: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const comicResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const storiesResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {} as any,
      characters: {} as any,
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {} as any,
      type: '',
    },
  ]);

const storyResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {
        resourceURI: 'apiUrl/comics/1',
        name: 'Spiderman Comic',
      },
      type: '',
    },
  ]);

export {
  createGenericApiResponse,
  emptyResponse,
  charactersResponse,
  characterResponse,
  comicsResponse,
  comicResponse,
  storiesResponse,
  storyResponse,
};
