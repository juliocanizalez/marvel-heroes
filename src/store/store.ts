/* eslint-disable no-underscore-dangle */
import { createStore, combineReducers, StoreEnhancer } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import SearchReducer from '../reducers/search/SearchReducer';
import SearchComicReducer from '../reducers/searchComic/SearchComicReducer';
import ViewItemReducer from '../reducers/viewItem/ViewItemReducer';
import LocalItemsReducer from '../reducers/localItems/LocalItemsReducer';

type WindowWithDevTools = Window & {
  __REDUX_DEVTOOLS_EXTENSION__: () => StoreEnhancer<unknown, {}>;
};

const isReduxDevtoolsExtenstionExist = (
  arg: Window | WindowWithDevTools,
): arg is WindowWithDevTools => '__REDUX_DEVTOOLS_EXTENSION__' in arg;

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['localItems'],
};

const rootReducer = combineReducers({
  search: SearchReducer,
  searchComic: SearchComicReducer,
  viewItem: ViewItemReducer,
  localItems: LocalItemsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = createStore(
  persistedReducer,
  isReduxDevtoolsExtenstionExist(window) ? window.__REDUX_DEVTOOLS_EXTENSION__() : undefined,
);

export const persistor = persistStore(store);

export type IRootState = ReturnType<typeof rootReducer>;

export default store;
