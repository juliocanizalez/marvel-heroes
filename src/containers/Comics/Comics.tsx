/* eslint-disable */

import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import queryString from 'query-string';
import { useHistory } from 'react-router-dom';
import { debounce } from 'lodash';
import { faBookmark as faBookmarkSolid } from '@fortawesome/free-solid-svg-icons';
import { faBookmark as faBookmarkRegular } from '@fortawesome/free-regular-svg-icons';

import { IRootState } from '../../store/store';
import IGenericApiResponse from '../../interfaces/IGenericApiResponse';
import IComic from '../../interfaces/IComic';
import Card from '../../components/Card/Card';
import Loading from '../../components/Loading/Loading';
import Pagination from '../../components/Pagination/Pagination';
import {
  comicReset,
  setComicAllParams,
  setComicAsyncContent,
  setComicBaseUrl,
  setComicFormat,
  setComicTitle,
} from '../../actions/SearchComic';
import { addBookmark, hideLocalItem, removeBookmark } from '../../actions/LocalItems';
import { getQueryVariable, shortenText, imagePlaceHolder } from '../../utils/utils';
import { get } from '../../api/FetchInfo';
import FormatOptions from '../../mocks/FormatOptions';
import './Comics.scss';

const Comics: React.FC = () => {
  enum QuerysParams {
    Page = 'page',
    Format = 'format',
    Title = 'title',
  }

  const history = useHistory();
  const dispatch = useDispatch();
  const {
    url,
    limit,
    format,
    title,
    loading,
    error,
    data: genericResponse,
  } = useSelector((state: IRootState) => state.searchComic);
  const { hiddenItems, bookmarks } = useSelector((state: IRootState) => state.localItems);

  const { data } = genericResponse ?? {};
  const { results } = data ?? {};
  const [comics, setComics] = useState<IComic[] | null>(null);

  const changeUrlParams = (newPage: number, newFormat: string, newTitle: string) => {
    history.push(
      `?${queryString.stringify({
        page: newPage,
        format: newFormat === '' ? undefined : newFormat,
        title: newTitle === '' ? undefined : newTitle,
      })}`,
    );
  };

  const handleChangePage = (newPage: number) => {
    window.scrollTo({ top: 300 });
    changeUrlParams(newPage, format, title);
  };

  const debouncedTitleChange = useCallback(
    debounce((title: string, format: string) => changeUrlParams(1, format, title), 1000),
    [],
  );

  const handleTitleChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setComicTitle(target.value));
    debouncedTitleChange(target.value, format);
  };

  const handleFormatChange = ({ target }: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(setComicFormat(target.value));

    changeUrlParams(1, target.value, title);
  };

  const handleSubmit = (e: React.ChangeEvent<HTMLFormElement>) => e.preventDefault();

  const handleViewMore = (id: number) => history.push(`comics/${id}`);

  const handleHideItem = (id: number) => dispatch(hideLocalItem({ id, type: 'COMIC' }));

  const handleAddBookmark = (id: number) => dispatch(addBookmark({ id, type: 'COMIC' }));

  const handleRemoveBookmark = (id: number) => dispatch(removeBookmark({ id, type: 'COMIC' }));

  useEffect(() => {
    dispatch(setComicBaseUrl(`${process.env.REACT_APP_API_URL}v1/public/comics`));

    return () => {
      dispatch(comicReset());
    };
  }, []);

  useEffect(() => {
    const newPage = getQueryVariable(QuerysParams.Page) ? +getQueryVariable(QuerysParams.Page) : 1;

    const newFormat = getQueryVariable(QuerysParams.Format)
      ? decodeURI(getQueryVariable(QuerysParams.Format))
      : '';

    const newTitle = getQueryVariable(QuerysParams.Title)
      ? decodeURI(getQueryVariable(QuerysParams.Title))
      : '';

    const newOffset = Math.ceil((newPage - 1) * limit);

    dispatch(setComicAllParams(newOffset, newFormat, newTitle));
  }, [history.location, limit]);

  useEffect(() => {
    if (url) {
      dispatch(setComicAsyncContent(true, '', null));
      get<IGenericApiResponse<IComic>>(url)
        .then((res) => {
          dispatch(setComicAsyncContent(false, '', res));
        })
        .catch((err) => {
          dispatch(setComicAsyncContent(false, 'Could not load comics', null));
        });
    }
  }, [url]);

  useEffect(() => {
    if (results) {
      const hiddenComics = hiddenItems.filter((item) => item.type === 'COMIC');
      setComics(
        results.filter((char) => {
          const hiddenComicFilter = hiddenComics.filter((item) => item.id === char.id);
          if (hiddenComicFilter.length > 0) {
            return false;
          }
          return true;
        }),
      );
    }
  }, [results, hiddenItems]);

  return (
    <div className='comics'>
      <div className='comics-page__title'>
        <h1>Comics</h1>
      </div>
      <div className='comics-search'>
        <form onSubmit={handleSubmit}>
          <h3 className='comics-search__title'>Search your comic</h3>
          <div className='search'>
            <div className='search-header'>Title</div>
            <div className='search-value'>
              <input
                type='text'
                name='title'
                placeholder="Comic's title"
                autoComplete='off'
                className='search-input'
                value={title}
                onChange={handleTitleChange}
              />
            </div>
            <div className='search-value'>
              <select className='search-value__select' value={format} onChange={handleFormatChange}>
                <option value=''>Select a format</option>
                {FormatOptions.map((format) => (
                  <option key={format.value} value={format.value}>
                    {shortenText(format.name, 20)}
                  </option>
                ))}
              </select>
            </div>
          </div>
        </form>
      </div>
      {loading && <Loading />}
      {error && <h2 className='error-message'>Could not load comics 😓</h2>}
      <div className='cards'>
        <div className='cards-content'>
          {!loading &&
            comics &&
            comics.map((comic) => {
              const inBookmark = bookmarks.find(
                (item) => item.type === 'COMIC' && item.id === comic.id,
              );
              return (
                <Card
                  key={comic.id}
                  id={comic.id}
                  name={comic.title}
                  description={comic.description ?? ''}
                  bookmarkClassName={
                    inBookmark
                      ? 'btn btn-action btn-bookmark bookmark-selected'
                      : 'btn btn-action btn-bookmark'
                  }
                  bookmarkIcon={inBookmark ? faBookmarkSolid : faBookmarkRegular}
                  handleViewMore={handleViewMore}
                  handleHideItem={handleHideItem}
                  handleBookmarkAction={inBookmark ? handleRemoveBookmark : handleAddBookmark}
                  imageUrl={
                    comic.thumbnail
                      ? `${comic.thumbnail.path}/portrait_uncanny.${comic.thumbnail.extension}`
                      : imagePlaceHolder
                  }
                />
              );
            })}
        </div>
      </div>
      {results && results.length <= 0 && <h2 className='error-message'>Comics not found 😮</h2>}
      {!loading && results && results.length > 0 && comics && comics.length <= 0 && (
        <h2 className='error-message'>Comics on this page are hidden 🤐</h2>
      )}
      {!loading && !error && results && genericResponse && results.length > 0 && (
        <Pagination genericResponse={genericResponse} onChange={handleChangePage} />
      )}
    </div>
  );
};

export default Comics;
