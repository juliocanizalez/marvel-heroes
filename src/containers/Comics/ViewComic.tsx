import React, { useEffect, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faChevronLeft,
  faEyeSlash,
  faBookmark as faBookmarkSolid,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons';
import { faBookmark as faBookmarkRegular } from '@fortawesome/free-regular-svg-icons';
import queryString from 'query-string';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';

import { IRootState } from '../../store/store';
import { setViewItemAsyncContent } from '../../actions/ViewItem';
import { get } from '../../api/FetchInfo';
import IGenericApiResponse from '../../interfaces/IGenericApiResponse';
import IComic from '../../interfaces/IComic';
import Loading from '../../components/Loading/Loading';
import CustomOrderedList from '../../components/CustomOrderedList/CustomOrderedList';
import RouteNames from '../../routes/RoutePaths';
import { imagePlaceHolder } from '../../utils/utils';
import { addBookmark, hideLocalItem, removeBookmark } from '../../actions/LocalItems';
import './Comics.scss';

const ViewComic: React.FC = () => {
  interface PathParams {
    idComic: string;
  }

  const history = useHistory();
  const dispatch = useDispatch();
  const { idComic } = useParams<PathParams>();
  const {
    loading,
    error,
    data: genericResponse,
  } = useSelector((state: IRootState) => state.viewItem);
  const { hiddenItems, bookmarks } = useSelector((state: IRootState) => state.localItems);

  const { data } = genericResponse ?? {};
  const { results } = data ?? {};
  const [comic, setComic] = useState<IComic | null>(null);
  const [inBookmark, setInBookmark] = useState(
    !!bookmarks.find((item) => item.type === 'COMIC' && item.id === parseFloat(idComic)),
  );
  const [bookmarkIcon, setBookmarkIcon] = useState<IconDefinition>(
    inBookmark ? faBookmarkSolid : faBookmarkRegular,
  );
  const [hidden] = useState(
    !!hiddenItems.find((item) => item.type === 'COMIC' && item.id === parseFloat(idComic)),
  );

  const handleBack = () => {
    if (history.length <= 2) history.push(RouteNames.HOME);
    else history.goBack();
  };

  const handleViewChar = (id: string) => history.push(`/characters/${id}`);

  const handleViewStory = (id: string) => history.push(`/stories/${id}`);

  const handleAddBookmark = (id: number) => {
    dispatch(addBookmark({ id, type: 'COMIC' }));
    setBookmarkIcon(faBookmarkSolid);
    setInBookmark(true);
  };

  const handleRemoveBookmark = (id: number) => {
    dispatch(removeBookmark({ id, type: 'COMIC' }));
    setBookmarkIcon(faBookmarkRegular);
    setInBookmark(false);
  };

  const handleHideItem = (id: number) => {
    dispatch(hideLocalItem({ id, type: 'COMIC' }));
    handleBack();
  };

  useEffect(() => {
    if (results) {
      if (!('comics' in results[0])) {
        setComic(results[0] as unknown as IComic);
      }
    }
  }, [results]);

  useEffect(() => {
    dispatch(setViewItemAsyncContent(true, '', null));
    const url = `${
      process.env.REACT_APP_API_URL
    }v1/public/comics/${idComic}?${queryString.stringify({
      apikey: process.env.REACT_APP_PUBLIC_KEY,
    })}`;
    get<IGenericApiResponse<IComic>>(url)
      .then((res) => {
        dispatch(setViewItemAsyncContent(false, '', res));
      })
      .catch(() => {
        dispatch(setViewItemAsyncContent(false, 'Could not load comic', null));
      });
  }, []);

  return (
    <div className='main-content'>
      {loading && <Loading />}
      {error && <h2 className='error-message'>Could not load comic ðŸ˜“</h2>}
      {!loading && !error && comic && (
        <>
          <div className='image-header-div'>
            <button type='button' className='back-button' onClick={handleBack}>
              <FontAwesomeIcon icon={faChevronLeft} />
              {'\u00A0'}
              Back
            </button>
            <h1 className='image-header-title'>{comic.title}</h1>
          </div>
          <div className='comic-description'>
            <div
              className='comic-image'
              style={{
                backgroundImage: `url('${
                  comic.thumbnail
                    ? `${comic.thumbnail.path}/portrait_uncanny.${comic.thumbnail.extension}`
                    : imagePlaceHolder
                }')`,
              }}
            />
            <div className='comic-info'>
              <p className='comic-description-p'>{comic.description}</p>
              {comic.urls && comic.urls.length > 0 && (
                <>
                  <p className='comic-extra-title'>{`Learn more about ${comic.title}`}</p>
                  <div className='comic-extra-div'>
                    {comic.urls.map((res) => (
                      <a
                        className='comic-extra-button'
                        target='_blank'
                        rel='noopener noreferrer'
                        key={res.type}
                        href={res.url}
                      >
                        {res.type}
                      </a>
                    ))}
                  </div>
                </>
              )}
            </div>
          </div>
          <div className='actions-title-div'>
            <p className='actions-title'>Actions:</p>
          </div>
          <div className='actions-menu'>
            <button
              type='button'
              className='action-btn add-bookmark'
              onClick={() => {
                if (inBookmark) handleRemoveBookmark(parseFloat(idComic));
                else handleAddBookmark(parseFloat(idComic));
              }}
            >
              <FontAwesomeIcon icon={bookmarkIcon} />
              {'\u00A0'}
              {inBookmark ? 'Remove bookmark' : 'Add bookmark'}
            </button>
            {!hidden && (
              <button
                type='button'
                className='action-btn hide-item'
                onClick={() => {
                  handleHideItem(parseFloat(idComic));
                }}
              >
                <FontAwesomeIcon icon={faEyeSlash} />
                {'\u00A0'}
                Hide item
              </button>
            )}
          </div>
          <div className='related'>
            <div className='related-items'>
              <h3 className='related-items-title'>Characters</h3>
              {comic.characters.returned > 0 ? (
                <CustomOrderedList items={comic.characters.items} onClick={handleViewChar} />
              ) : (
                <p className='not-found'>No characters found ðŸ¤”</p>
              )}
            </div>
            <div className='related-items-div fade-in'>
              <h3 className='related-items-title'>Stories</h3>
              {comic.stories.returned > 0 ? (
                <CustomOrderedList items={comic.stories.items} onClick={handleViewStory} />
              ) : (
                <p className='not-found'>No stories found ðŸ¤”</p>
              )}
            </div>
          </div>
        </>
      )}
    </div>
  );
};

export default ViewComic;
