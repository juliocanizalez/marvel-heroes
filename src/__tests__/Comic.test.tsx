/* eslint-disable arrow-body-style */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-unused-vars */
/* eslint-disable import/extensions */
/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';

import store from '../store/store';
import server from '../mocks/server';
import { emptyResponse } from '../mocks/testData';
import Comics from '../containers/Comics/Comics';

describe('Test on Comics', () => {
  test('Should render component correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
      expect(container).toMatchSnapshot();
    });
  });

  test('should show typed title on input', async () => {
    const search = 'x-men';
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    userEvent.type(screen.getByPlaceholderText(/Comic's title/i), search);
    expect((screen.getByPlaceholderText(/Comic's title/i) as HTMLInputElement).value).toBe(search);
  });

  test('should add/remove bookmark', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
    });

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).not.toBe(null);

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).toBe(null);
  });

  test('should hide card and show hidden comics message', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
    });

    userEvent.click(container.querySelector('.btn-hide') as Element);

    expect(container.querySelector('.cards')).toBeNull();
    expect(screen.getByText(/Comics on this page are hidden 🤐/i)).toBeInTheDocument();
  });

  test('should handle no results', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(emptyResponse()));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Comics not found 😮/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
      expect(container.querySelector('.pagination-options')).toBeNull();
    });
  });

  test('should handle api error', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics`, (req, res, ctx) => {
        return res(ctx.status(404));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Comics />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Could not load comics 😓/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
      expect(container.querySelector('.pagination-options')).toBeNull();
    });
  });
});
