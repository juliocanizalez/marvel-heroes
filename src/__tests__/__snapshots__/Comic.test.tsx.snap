// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Comics Should render component correctly 1`] = `
<div>
  <div
    class="comics"
  >
    <div
      class="comics-page__title"
    >
      <h1>
        Comics
      </h1>
    </div>
    <div
      class="comics-search"
    >
      <form>
        <h3
          class="comics-search__title"
        >
          Search your comic
        </h3>
        <div
          class="search"
        >
          <div
            class="search-header"
          >
            Title
          </div>
          <div
            class="search-value"
          >
            <input
              autocomplete="off"
              class="search-input"
              name="title"
              placeholder="Comic's title"
              type="text"
              value=""
            />
          </div>
          <div
            class="search-value"
          >
            <select
              class="search-value__select"
            >
              <option
                value=""
              >
                Select a format
              </option>
              <option
                value="comic"
              >
                Comic
              </option>
              <option
                value="magazine"
              >
                Magazine
              </option>
              <option
                value="trade paperback"
              >
                Trade Paperback
              </option>
              <option
                value="hardcover"
              >
                Hardcover
              </option>
              <option
                value="digest"
              >
                Digest
              </option>
              <option
                value="graphic novel"
              >
                Graphic Novel
              </option>
              <option
                value="digital comic"
              >
                Digital Comic
              </option>
              <option
                value="infinite comic"
              >
                Infinite Comic
              </option>
            </select>
          </div>
        </div>
      </form>
    </div>
    
    <div
      class="cards"
    >
      <div
        class="cards-content"
      />
    </div>
    <div
      class="pagination-options"
    >
      <div
        class="multi-button"
      />
    </div>
    <div
      class="current-pagination-page"
    >
      <p>
        <b>
          Page: 
        </b>
        1 / 4130
      </p>
    </div>
  </div>
</div>
`;
